### PR TITLE
Fix build on non x86 archs.

### PR DESCRIPTION
**Pull request by devnexen**
Discard libfusion on other architectures.

### CODE REVIEW

File: include/picotls/fusion.h [hunk 1]
```diff
@@ -91,6 +92,7 @@ extern ptls_aead_algorithm_t ptls_fusion_aes128gcm, ptls_fusion_aes256gcm;
  * Returns a boolean indicating if fusion can be used.
  */
 int ptls_fusion_is_supported_by_cpu(void);
+#endif
```
Comment (huitema) [inline include/picotls/fusion.h:96]: I think that you want to keep the declaration of `ptls_fusion_is_supported_by_cpu`, but modify `fusion.c` to return false for non x86 architectures.

**Review by huitema — COMMENTED** [review]
Makes sense, but please check the support of `ptls_fusion_is_supported_by_cpu`

**Review by huitema — CHANGES_REQUESTED** [review]
The way the macros are written, fusion would also be discarded when using the Visual Studio compiler. Please apply suggested changes.

File: include/picotls/fusion.h [hunk 2]
```diff
@@ -26,6 +26,7 @@
 extern "C" {
 #endif
 
+#if defined(__x86_64__) || defined(__amd64__)
 #include <stddef.h>
```
Comment (huitema) [inline include/picotls/fusion.h:30]: That does not work -- the variables __x86_64___ or __amd64__ are not defined by the VS compiler. What does work is:
```
#if defined(_WINDOWS64) || defined(__x86_64__) || defined(__amd64__)
```

File: t/ptlsbench.c [hunk 3]
```diff
@@ -255,7 +255,7 @@ static ptls_bench_entry_t aead_list[] = {
     {"ptlsbcrypt", "aes128gcm", &ptls_bcrypt_aes128gcm, &ptls_bcrypt_sha256, 1},
     {"ptlsbcrypt", "aes256gcm", &ptls_bcrypt_aes256gcm, &ptls_bcrypt_sha384, 1},
 #endif
-#if !defined(_WINDOWS) || defined(_WINDOWS64)
+#if (defined(__x86_64__) || defined(__amd64__)) && (!defined(_WINDOWS) || defined(_WINDOWS64))
     {"fusion", "aes128gcm", &ptls_fusion_aes128gcm, &ptls_minicrypto_sha256, 1},
```
Comment (huitema) [inline t/ptlsbench.c:259]: That does not work -- the variables __x86_64___ or __amd64__ are not defined by the VS compiler. What does work is:
```
#if defined(_WINDOWS64) || defined(__x86_64__) || defined(__amd64__)
```

**Review by huitema — CHANGES_REQUESTED** [review]
Oops, Github ate my comments. Please fix the `defined()` tests.

File: include/picotls/fusion.h (hunk 1 above)
Comment (devnexen) [reply include/picotls/fusion.h:96]: I m not sure about this one as I see in the build the fusion part is only build for AMD/x86_64 ?

Comment (huitema) [reply include/picotls/fusion.h:96]: In fusion.c, do something like:

```
#if defined(_WINDOWS64) || defined(__x86_64__) || defined(__amd64__)
... current definitions of ptls_fusion_is_supported_by_cpu
#else 
int ptls_fusion_is_supported_by_cpu(void)
{
    return(0);
}
#endif
```

That will also solve some potential compilation warnings for those compilers that do not like empty code files.

Comment (huitema) [reply include/picotls/fusion.h:96]: @devnexen  Sorry, I am wrong and you are right. Fusion.c will not be compiled at all if the architecture in not x86, so you do not need to provide a definition for `ptls_fusion_is_supported_by_cpu`

**Review by huitema — APPROVED** [review]

**Review by kazuho — COMMENTED** [review]
Thank you for working on to resolve this problem.

I tend to think that there are two principled approaches that we could take, and that ATM this PR is kind of in the middle.

One principle that we can adopt is that fusion is specific to x86-64, and that it should not be used on other platforms. This is much like our support for the openssl backend. lib/openssl.c is not expected to compile when OpenSSL is not installed. lib/openssl.h is not expected to be used when OpenSSL is not installed.

The other principle that we might want to adopt is that fusion as a backend is accessible on all platforms, even when it cannot be used. If that is the case, fusion.h should compile on all platforms _and_ in order to avoid link errors, all the functions of fusion have to exist on all platforms. Users will be required to call `ptls_fusion_is_supported_by_cpu` to see if it can use those functions.

Looking at the discussion on this PR, it seems that the current state is in the middle of the two. `#if`are added to fusion.h to let it compile on architectures other than x86-64, but fusion.c is assumed to not be compiled. As stated above, that leads to link error. I do not think that is the state that we want to land.

IMO, even though we might add support for archs other than x86-64 in the future, I tend to think that fusion is specific to x86-64. Based on that, I tend to think that we should adopt the former approach.

That is, make no changes to fusion.h or fusion.c, but add architecture-specific `#if` to the code that uses fusion.

WDYT?

**Comment by huitema** [comment]
@kazuho, you suggest to," make no changes to fusion.h or fusion.c, but add architecture-specific #if to the code that uses fusion". This does indeed work, but for an application that supports multiple platforms it requires writing something like:

```
#if (some test of platform support).
#include "fusion.h"
#endif 
```

Then you need to do it at multiple places in the code, such as for example before calling `is fusion supported`. This is what I ended up doing in `ptlsbench.ct`.

By the way, I believe that applications are not supposed to call the inner fusion API. They just have to import the function `ptls_fusion_is_supported_by_cpu` and the declaration of the Cipher and AEAD algorithms:

```
extern ptls_cipher_algorithm_t ptls_fusion_aes128ctr, ptls_fusion_aes256ctr;
extern ptls_aead_algorithm_t ptls_fusion_aes128gcm, ptls_fusion_aes256gcm;
```

For run time decisions on not supported platforms, it would be sufficient to have a definition of the function that `ptls_fusion_is_supported_by_cpu` and some proforma declaration of the algorithms.

**Review by blaquezzz — APPROVED** [review]

**Comment by kazuho** [comment]
Sorry for the delay,

@huitema Maybe the correct thing to do is provide a macro like `PTLS_HAVE_FUSION`. Then, applications can use `#if` to compile code that use fusion. They have to also call `ptls_fusion_issupported_by_cpu` as runtime checks are also necessary, but I think that's inevitable.

**Comment by huitema** [comment]
@kazuho There are two issues: can the code compile, and can the code run. The code in `fusion.c` can only compile if the compiler does not generate an error when encountering extension instructions. Then, the functions are only useful if the particular CPU can provide efficient support for these instructions. One is a compile time issue, which could be treated using configuration flags in CMake, or the equivalent in Visual Studio projects -- this is how we currently address OpenSSL. The other is an execution time issue, and that indeed requires calling `ptls_fusion_issupported_by_cpu`. Having a macro like `PTLS_HAVE_FUSION` would simplify the compile issues, because it could be set by CMake or VS.

**Comment by kazuho** [comment]
Thank you for the PR.

As discussed here, it would be more consistent to _not_ compile fusion on platforms other than x86-64, as we do for openssl backend. Closing.

**Pull request closed without merging**